dffml/tuner/random_search.py

```diff
@@ -0,0 +1,127 @@
+from typing import Union, Dict, Any
+import itertools
+import logging
+import random
+
+from ..base import (
+    config,
+    field,
+)
+from ..high_level.ml import train, score
+from .tuner import Tuner, TunerContext
+from ..util.entrypoint import entrypoint
+from ..source.source import BaseSource, Record
+from ..accuracy.accuracy import AccuracyScorer, AccuracyContext
+from ..model import ModelContext
+from ..feature.feature import Feature
+
+
+@config
+class RandomSearchConfig:
+    parameters: dict = field("Parameters to be optimized")
+    objective: str = field(
+        "How to optimize the given scorer. Values are min/max", default="max"
+    )
+    trials: int = field("Number of random configurations to try.", default=20)
+
+
+class RandomSearchContext(TunerContext):
+    """
+    Parameter Grid Tuner
+    """
+
+    async def optimize(
+        self,
+        model: ModelContext,
+        feature: Feature,
+        accuracy_scorer: Union[AccuracyScorer, AccuracyContext],
+        train_data: Union[BaseSource, Record, Dict[str, Any]],
+        test_data: Union[BaseSource, Record, Dict[str, Any]],
+    ):
+        """
+        Method to optimize hyperparameters by parameter grid.
+        Uses a grid of hyperparameters in the form of a dictionary present in config,
+        Trains each permutation of the grid of parameters and compares accuracy.
+        Sets model to the best parameters and returns highest accuracy.
+
+        Parameters
+        ----------
+        model : ModelContext
+            The Model which needs to be used.
+
+        feature : Feature
+            The Target feature in the data.
+
+        accuracy_scorer: AccuracyContext
+            The accuracy scorer that needs to be used.
+
+        train_data: SourcesContext
+            The train_data to train models on with the hyperparameters provided.
+
+        test_data : SourcesContext
+            The test_data to score against and optimize hyperparameters.
+
+        Returns
+        -------
+        float
+            The highest score value
+        """
+
+        if self.parent.config.objective == "min":
+            highest_acc = float("inf")
+        elif self.parent.config.objective == "max":
+            highest_acc = -1
+        else:
+            raise NotImplementedError('Objective must be either "min" or "max".')
+
+        best_config = dict()
+        logging.info(
+            f"Optimizing model with parameter grid: {self.parent.config.parameters}"
+        )
+
+        names = list(self.parent.config.parameters.keys())
+        logging.info(names)
+
+        with model.parent.config.no_enforce_immutable():
+            for _ in range(self.parent.config.trials):
+                combination = []
+                for pvs in self.parent.config.parameters.values():
```

Comment:
what does `.values()` get here?

Reply:
It's the values for the search space of the parameter. i.e [1,10,100] for n_iterations.